File: DesignPatternPractice_Python/TemplateMethodPattern.py
from abc import ABCMeta, abstractmethod
from typing import Type
from functools import wraps


class TemplateMethodPattern:
    @staticmethod
    def run():
        class Counter:
            def __init__(self):
                self.i = 1

        def counter_decorator(counter: Counter):
            def inner_wrapper(func):
                @wraps(func)
                def decorated_func(*args, **kwargs):
                    print(f"Step {counter.i}: ")
                    counter.i += 1
                    return func(*args, **kwargs)

                return decorated_func

            return inner_wrapper

        class AbstractDish(metaclass=ABCMeta):
            @abstractmethod
            def prepare(self):
                pass

            @abstractmethod
            def cook(self):
                pass

            @abstractmethod
            def apply_cumin(self):
                pass

            @abstractmethod
            def eat(self):
                pass

            @abstractmethod
            def clear(self):
                pass

            # Note: this is Hook Method, i.e., the method that have impacts on the template method
            def allow_apply_cumin(self):  # default, apply cumin
                return True

            def happy_time(self):
                self.prepare()
                self.cook()
                if self.allow_apply_cumin():
                    self.eat()
                self.eat()
                self.clear()

        class GrillBeef(AbstractDish):
            counter = Counter()

            def __init__(self):
                self.allow_apply_cumin_flag = False

            def set_allow_apply_cumin_flag(self, flag: bool):  # Customizable
                self.allow_apply_cumin_flag = flag

            @counter_decorator(counter)
            def prepare(self):
                print("Prepare grill beef")

            @counter_decorator(counter)
            def cook(self):
                print("Cook grill beef")

            @counter_decorator(counter)
            def apply_cumin(self):
                print("Add some cumin to beef")

            @counter_decorator(counter)
            def eat(self):
                print("Eat grill beef")

            @counter_decorator(counter)
            def clear(self):
                print("Clear grill beef\n")

            def allow_apply_cumin(self):
                return self.allow_apply_cumin_flag

        class GrillLamb(AbstractDish):
            counter = Counter()

            @counter_decorator(counter)
            def prepare(self):
                print("Prepare grill lamb")

            @counter_decorator(counter)
            def cook(self):
                print("Cook grill lamb")

            @counter_decorator(counter)
            def apply_cumin(self):
                print("Add some cumin to lamb")

            @counter_decorator(counter)
            def eat(self):
                print("Eat grill lamb")

            @counter_decorator(counter)
            def clear(self):
                print("Clear grill lamb\n")

            def allow_apply_cumin(self):  # Always apply cumin for lamb
                return True

        grill_beef = GrillBeef()
        grill_beef.happy_time()
        grill_lamb = GrillLamb()
        grill_lamb.happy_time()

File: DesignPatternPractice_CSharp/Source/TemplateMethodPattern.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DesignPatternPractice.Source
{
    internal class TemplateMethodPattern
    {
        internal abstract class AbstractDish
        {
            // Use protect to follow LoD
            protected abstract void Prepare();
            protected abstract void Cook();
            protected abstract void ApplyCumin();
            protected abstract void Eat();
            protected abstract void Clear();

            // Note: this is Hook Method, i.e., the method that have impacts on the template method
            protected virtual bool AllowApplyCumin() => true;  // default, apply cumin

            internal void HappyTime()
            {
                int i = 0;
                Console.WriteLine($"Step {++i}: ");
                Prepare();
                Console.WriteLine($"Step {++i}: ");
                Cook();
                if (AllowApplyCumin())
                {
                    Console.WriteLine($"Step {++i}: ");
                    ApplyCumin();
                }
                Console.WriteLine($"Step {++i}: ");
                Eat();
                Console.WriteLine($"Step {++i}: ");
                Clear();
            }
        }

        internal class GrillBeef : AbstractDish
        {
            private bool allowApplyCuminFlag = false;
            protected override void Prepare() => Console.WriteLine("Prepare grill beef");
            protected override void Cook() => Console.WriteLine("Cook grill beef");
            protected override void ApplyCumin() => Console.WriteLine("Add some cumin to beef");
            protected override void Eat() => Console.WriteLine("Eat grill beef");
            protected override void Clear() => Console.WriteLine("Clear grill beef\n");

            public void SetAllowApplyCuminFlag(bool flag) => allowApplyCuminFlag = flag;  // Customizable
            protected override bool AllowApplyCumin() => allowApplyCuminFlag;
        }

        internal class GrillLamb : AbstractDish
        {
            protected override void Prepare() => Console.WriteLine("Prepare grill lamb");
            protected override void Cook() => Console.WriteLine("Cook grill lamb");
            protected override void ApplyCumin() => Console.WriteLine("Add some cumin to lamb");
            protected override void Eat() => Console.WriteLine("Eat grill lamb");
            protected override void Clear() => Console.WriteLine("Clear grill lamb\n");

            protected override bool AllowApplyCumin() => true;  // Always apply cumin for lamb
        }

        public static void Run()
        {
            AbstractDish grillBeef = new GrillBeef();
            grillBeef.HappyTime();
            AbstractDish grillLamb = new GrillLamb();
            grillLamb.HappyTime();
        }
    }
}
